### ISSUE TITLE
idea: Separately learn whether there can potentially be a color change on a particular layer

### ISSUE THREAD

**Issue by rparrett** [hvoss-techfak/AutoForge#15]
The current method of pruning to reduce filament swaps is pretty effective, but I wonder if adding another parameter that encodes "should we even swap" might help bias the optimization against pointless or regressive swaps.

The idea would be for the current `global_logits` to represent "the color if a swap occurs" and to add a new param of the same shape repesenting a boolean "can a swap occur on this layer?"

Curious to know your thoughts on the subject. This is just a naive idea that I may experiment with later.

It also occurred to me that some sort of pruning could potentially be performed just based on simple math on the TD, noting the number of consecutive layers a color can be used before it becomes opaque, though I don't think that would be very useful in practice.

**Comment by hvoss-techfak** [comment]
Thanks for the ideas. I actually tried splitting the `global_logits` into a `learned color` (which color should be set) and `layer_swap_logit` (where should this swap occur) to limit the amounts of color swaps from the beginning and allowing for more control over the learning process as part of my exploration into pruning. Sadly, this gave me very bad results as we are introducing a lot of local minima in which the learning process gets stuck. 

I actually also tried to use a genetic algorithm to optimize these two variables but the search space is simply far too big to be learned efficiently with non-gradient methods. I still think there is a lot of room for improvement on the optimization part. Maybe something like intermittent layer pruning would help the learning process or some entirely different method. If you find out anything please let me know


For the pruning, the solution right now actually works pretty good as we just leverage the loss we already have. Maybe we could introduce a penalty during the optimization to penalize any colors that exceed their TD and become fully opaque. 
Maybe we could actually do what you suggested as a first step before all of the other pruning, to simply remove fully opaque layers. I have a bit more time this weekend and maybe try to see if that works.
